nilearn/regions/region_extractor.py

```diff
@@ -0,0 +1,480 @@
+from .._utils import check_niimg, check_niimg_3d, check_niimg_4d
+from .._utils.extmath import fast_abs_percentile
+from .._utils.class_inspect import get_params
+from ..image import iter_img, new_img_like
+from ..image.image import _smooth_array
+from .._utils.niimg_conversions import concat_niimgs
+from .._utils.compat import _basestring
+from ..externals.skimage import peak_local_max, random_walker
+
+
+def estimate_apply_threshold_to_maps(maps_img, mask_img=None, parameters=None,
+                                     threshold='auto',
+                                     thresholding_strategy='percentile'):
+    """ A function which keeps the most prominent regions of the maps. Can also
+    be denoted as foreground objects/regions extraction.
+
+    Parameters
+    ----------
+    maps_img: a 3D/4D Nifti like image/object
+        a path or filename or an image consists of statistical maps.
+
+    mask_img: a Nifti like image/object, default is None
+        a path or filename or a mask image. To make sure that we keep regions
+        which are within the brain volume.
+        If given, maps are masked with this image or if not given, maps are
+        masked automatically estimated by a `NiftiMasker`.
+
+    parameters: list of parameters, default is None
+        If given, parameters are passed to `filter_and_mask` along with
+        the provided input mask image.
+
+    threshold: a float (t-value, ratio), 'auto', default is 'auto'
+        a value used to threshold the maps.
+
+        If float, a t-statistic threshold value of maps, the intensities
+        which are survived above this value are kept. Most suitable if user
+        knows exactly the values to keep.
+
+        If float, a ratio value between [0. and 1.], is used to multiply
+        with the total size of voxels (ratio * n_voxels = some_voxels).
+        The most intense voxels (some_voxels) which are survived will be kept.
+        Most suitable for the case, if the concept of threshold should be
+        according to total count of voxels inside the brain volume.
+
+        If default 'auto', pre-defined ratio 0.8 is used, which means 80%
+        of the voxels which are survived are kept.
+
+    thresholding_strategy: string {None, 'ratio_n_voxels', 'percentile'} \
+        default 'percentile'
+        a parameter to select between the type of determining a threshold value.
+        It can be either based on number of voxels or score of the data based
+        on a percentile.
+
+        If None, threshold value will not be determined.
+
+        If ratio_n_voxels, threshold value is estimated based on total size of the
+        voxels.
+
+        If percentile, threshold value is determined based on the percentage
+        score of the input data of the maps.
+
+    Returns
+    -------
+    threshold_maps_img: a Nifti like image/object
+        a thresholded image of the input.
+    """
+    maps_img = check_niimg(maps_img)
```

Comment:
`at_least_4d=True` will turn your 3D data into 4D and make the things easier for you.